### PR TITLE
Can Get Shell By Case Insensitive Name.

### LINKED ISSUE #12106: Allow case-insensitive tenant names in ShellHost
### Is your feature request related to a problem? Please describe.
In a mobile version, I created a page when a user can login to a specific tenant by providing the `tenant name`, `username`, `password`.  To identify which tenant the user is trying to access I user `_shellHost.TryGetSettings(tenantName, out settings)` which looks up the tenant by name. The issue here is if the user does not provide a case sensitive tenant name, the `ShellHost` does not find anything. 

### Describe the solution you'd like
It would be helpful if user case insensitive to find tenants. We can easily achieve this by storing the tenant key as lowercase and when we get/remove the tenant we will also use lowercase so no matter what the value is, we treat it the same. 

@jtkech is there an issue from making this change? I can push a PR if no issue with doing this.

For example the line [if (!_shellContexts.TryAdd(settings.Name, new ShellContext.PlaceHolder { Settings = settings }))] (https://github.com/OrchardCMS/OrchardCore/blob/51f0db7ac0d621fe3ecc415e707df58f81883c00/src/OrchardCore/OrchardCore/Shell/ShellHost.cs#L197) will become something like this instead `if (!_shellContexts.TryAdd(settings.Name.ToLower(), new ShellContext.PlaceHolder { Settings = settings }))`

Obviosly, we have to make the same change on `TryGetValue` and `TryRemove`
### Describe alternatives you've considered
I did the following

```
private ShellSettings GetSettings(string tenantName)
{
    if (_shellHost.TryGetSettings(tenantName, out ShellSettings settings))
    {
        return settings;
    }

    var shells = _shellHost.ListShellContexts();
    var tenant = shells.FirstOrDefault(x => x.Settings != null && x.Settings.Name.Equals(tenantName, StringComparison.OrdinalIgnoreCase));

    if (tenant != null)
    {
        return tenant.Settings;
    }

    return null;
}
```

### PR DESCRIPTION
**Pull request by CrestApps**
Fix #12106

@jtkech this PR replaces #12114 as per your request

### CODE REVIEW

File: src/OrchardCore/OrchardCore/Shell/ShellHost.cs
Comment (CrestApps) [reply src/OrchardCore/OrchardCore/Shell/ShellHost.cs:474]: So no need for the finalizer here?

File: src/OrchardCore/OrchardCore/Shell/ShellHost.cs
Comment (CrestApps) [reply src/OrchardCore/OrchardCore/Shell/ShellHost.cs:73]: Maybe we should move the second check in  `PreCreateAndRegisterShellsAsync` instead? So if that method is ever called again it won’t be executed

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:33]: Using `foundMixedShellSettings` asserts that we can get the same setting using case-insensitive (i.e, find using “test” when the name is “Test”) where as `foundShellSettings` just assets that it’s able to get the same exact name. Perhaps asserting that we can get the same exact this is not too important?

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:30]: Both shellSettings and ShellContext are now case insensitive. So I want to assert both get using case-insensitive name

File: src/OrchardCore/OrchardCore/Shell/ShellHost.cs
Comment (CrestApps) [reply src/OrchardCore/OrchardCore/Shell/ShellHost.cs:73]: Fixed

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs]: What about the updated test now?

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:25]: Done

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs]: My fault, I was using TryGetSettings instead of using TryGetShellContext. Please checkout the updated test

**Comment by jtkech** [comment]
@CrestApps 

Okay will work on it tomorrow.

Just did some little tweaks locally for the unit tests,

Will push it tomorrow to let you see them.

**Comment by hishamco** [comment]
@CrestApps why you closed #12114 it is not a good idea to close a PR and replace it unless there's a value. I had seen a lot of comments in the previous one during the review. Hope to consider this next time and thanks again for your contribution

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:11]: Move it down after the constructor

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:19]: Can we use `Theory` for the first two unit tests while they are similar?

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:56]: Remove extra lines

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:95]: add new line after

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:11]: Why `ShellHost` static?

**Comment by CrestApps** [comment]
I did that based on the request from @jtkech

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:83]: You can simplify by introducing `GetOrCreateShellContext()` as private method

**Review by hishamco — CHANGES_REQUESTED** [review]
I will leave the Shell related stuff to @jtkech, just a little notes about the unit tests

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:19]: @hishamco  Can you please explain why you are requesting to use theory? It's my understanding that `[Thoery]` would be used when there are parameters involved and `[Fact]` where there are no parameters.

Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs]: I think I understood your request. Please check the updated tests

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:69]: I'm not sure why we need this, you can supply various casing values in `Theory` args

Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:69]: I thought this way we only provide one variable and the logic will compare it to another variant. This way we don’t have to provide to parameters. Is there a reason we should provide two parameters instead?

Comment (hishamco) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:69]: Adding values with different cases will be more obvious to everyone instead of adding logic that doing something we don't know until we dig into the test

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:34]:
```csharp
Assert.NotEqual(shellName, anotherVariantOfShellName);
```

This will never fail while the values are different. Please let us know which values should be tested?

Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:34]: It will fail if the user accident provided equal parameters in the InlineData. I would leave it just in case someone made a change to the test. If both parameters are equal the test will pass but won't really test anything.

Comment (jtkech) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:34]: Cool, you are testing the test ;)

Comment (CrestApps) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:34]: Yup

**Comment by jtkech** [comment]
@hishamco

I already said to @CrestApps that I will update the tests with some suggestions, so maybe was too early to review them. The goal being to keep this PR as clean as possible because it acts on a sensitive  area.

> Just did some little tweaks locally for the unit tests,
> Will push it tomorrow to let you see them.

But that's okay as we still only have 2 files changed, so the PR is still "clean" ;)

I will do my suggestions soon.

Anyway the main point is to make `_shellContexts` and `_shellSettings` dictionaries case insensitive, I just need to think about it a little more and if I still think it is safe I will merge it (and if unit tests still pass).

Otherwise will wait for more feedback.

**Comment by jtkech** [comment]
@CrestApps 

I pushed my suggestions, feel free to revert or update what you don't like ;)

File: test/OrchardCore.Tests/Shell/ShellHostTests.cs
Comment (hishamco) [inline test/OrchardCore.Tests/Shell/ShellHostTests.cs:12]: @jtkech why you mark it as static? @CrestApps already did that at the beginning but I asked him to change it ;)

Comment (jtkech) [reply test/OrchardCore.Tests/Shell/ShellHostTests.cs:12]: Yes I saw this, because Site is a static thing so all can be static

**Comment by hishamco** [comment]
Seems the build is failed

**Comment by jtkech** [comment]
I saw it, I will fix it

Fail intermitently while running all tests in parallel and only for the Default tenant

**Pull request merged**
